### ISSUE TITLE
Optimize field fetching in FormattedDateTime

### ISSUE THREAD

**Issue by Manishearth** [unicode-org/icu4x#2143]
https://github.com/unicode-org/icu4x/pull/2138 moves us over to pre-extracting fields before constructing FormattedDateTime (etc). In many cases, not all fields are necessary, and fetching a field might be expensive, especially for things like week-of-day.

We should optimize this to not fetch fields that are not needed for the given set of options, probably with some kind of stored bitset on DTF.

**Comment by sffc** [comment]
This should be a benchmark-driven change, and it should look especially at non-Gregorian calendars.

**Comment by Manishearth** [comment]
Yeah. I think the main thing that will affect Gregorian users is week number / day of week fetching since those always involve computation.
